# Data Inspection

This notebook focuses on inspecting the structure and basic properties of the Titanic dataset before performing detailed data analysis.

The dataset used in this notebook is `Titanic-Dataset.csv`. Data inspection helps us understand the size, columns, data types, records, unique values, and possible roles of different variables.

Before building a Machine Learning model, it is important to inspect the dataset and understand its structure.

In [1]:
import pandas as pd

df = pd.read_csv("Titanic-Dataset.csv")

df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


## 1. Number of Rows

### Concept

The number of rows represents the total number of observations or records available in a dataset.

### Example

In the Titanic dataset, each row represents one passenger.

### Implementation

We use `df.shape[0]` to find the total number of rows.

### Output

The output displays the total number of passenger records.

### Interpretation

A larger number of rows generally means more observations are available for analysis.

### Insight

The number of rows helps us understand the overall size of the dataset.

### AI/ML Relevance

Machine Learning models require sufficient observations to learn meaningful patterns from the data.

In [2]:
print("Number of Rows:", df.shape[0])

Number of Rows: 891


## 2. Number of Columns

### Concept

The number of columns represents the total number of variables or features available in the dataset.

### Example

Columns such as `Age`, `Sex`, `Fare`, and `Pclass` describe different characteristics of Titanic passengers.

### Implementation

We use `df.shape[1]` to find the total number of columns.

### Output

The output displays the total number of variables in the dataset.

### Interpretation

Each column provides a different piece of information about a passenger.

### Insight

Understanding the number of columns helps us estimate the complexity of the dataset.

### AI/ML Relevance

The available columns can be used as input features or as a target variable for Machine Learning.

In [3]:
print("Number of Columns:", df.shape[1])

Number of Columns: 12


## 3. Dataset Dimensions

### Concept

Dataset dimensions describe the size of a dataset in terms of rows and columns.

### Example

A dataset with 891 rows and 12 columns contains 891 passenger records and 12 variables.

### Implementation

The `shape` attribute returns the number of rows followed by the number of columns.

### Output

The output displays the dataset dimensions in the format `(rows, columns)`.

### Interpretation

The first value represents the number of records, while the second value represents the number of variables.

### Insight

The Titanic dataset contains multiple passenger characteristics that can be inspected and analyzed.

### AI/ML Relevance

Dataset dimensions help us understand how many observations and potential features are available for model development.

In [4]:
print("Dataset Shape:", df.shape)

Dataset Shape: (891, 12)


## 4. Column Names

### Concept

Column names identify the variables available in a dataset.

### Example

In the Titanic dataset, columns such as `Age`, `Sex`, and `Fare` describe passenger characteristics.

### Implementation

The `columns` attribute is used to display all column names.

### Output

The output shows the complete list of columns.

### Interpretation

The column names help us understand what information is available in the dataset.

### Insight

Some columns may be useful as Machine Learning features, while others may act as identifiers or the target variable.

### AI/ML Relevance

Understanding column names is an important first step when selecting features for a Machine Learning model.

In [5]:
print("Column Names:")
print(df.columns.tolist())

Column Names:
['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']


## 5. Data Types

### Concept

A data type describes the kind of values stored in a column.

Common data types include integers, floating-point numbers, and text values.

### Example

`Age` usually contains numerical values, while `Sex` contains categorical text values.

### Implementation

The `dtypes` attribute displays the data type of every column.

### Output

The output shows the data type associated with each column.

### Interpretation

Data types help us determine whether a column should be treated as numerical, categorical, or another type.

### Insight

Incorrect data types may cause problems during data analysis or Machine Learning preprocessing.

### AI/ML Relevance

Understanding data types helps determine which preprocessing techniques, such as encoding or scaling, may be required.

In [6]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Cabin              str
Embarked           str
dtype: object

## 6. Dataset Index

### Concept

The index is used to identify the position of each row in a Pandas DataFrame.

### Example

The first passenger record may have index 0, the second record index 1, and so on.

### Implementation

The `index` attribute displays the row index information.

### Output

The output shows the range and structure of the dataset index.

### Interpretation

The index helps Pandas identify and access individual rows.

### Insight

The default index is generally not a meaningful passenger feature.

### AI/ML Relevance

The DataFrame index is usually not used as a Machine Learning feature unless it contains meaningful information.

In [7]:
print(df.index)

RangeIndex(start=0, stop=891, step=1)


## 7. First Records using head()

### Concept

The `head()` function displays the first few rows of a dataset.

### Example

By viewing the first passenger records, we can quickly inspect columns such as passenger class, name, sex, age, fare, and survival status.

### Implementation

By default, `head()` displays the first five rows.

### Output

The output shows the first five passenger records.

### Interpretation

This provides a quick overview of how the data is organized.

### Insight

The first records help us identify the types of information stored in different columns.

### AI/ML Relevance

A quick preview helps identify potential features and data issues before model development.

In [8]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


## 8. Last Records using tail()

### Concept

The `tail()` function displays the last few rows of a dataset.

### Example

Viewing the last passenger records helps confirm that the dataset contains records throughout the complete DataFrame.

### Implementation

By default, `tail()` displays the last five rows.

### Output

The output shows the final five passenger records.

### Interpretation

This helps inspect the end of the dataset and confirm its structure.

### Insight

Comparing the first and last records provides a broader understanding of how the dataset is organized.

### AI/ML Relevance

Inspecting different parts of the dataset can help identify unusual patterns or inconsistent data.

In [9]:
df.tail()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.00,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.00,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,NaN,Q


## 9. Random Records using sample()

### Concept

The `sample()` function selects random rows from a dataset.

### Example

Instead of always viewing the first or last passengers, we can inspect randomly selected passenger records.

### Implementation

We use `sample(5)` to display five random rows.

### Output

The output shows five randomly selected passenger records.

### Interpretation

Random sampling gives a more general view of the dataset.

### Insight

Random records can reveal different passenger categories and values present throughout the dataset.

### AI/ML Relevance

Random sampling is useful for quickly checking whether the data appears representative and consistent.

In [10]:
df.sample(5, random_state=42)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
709,710,1,3,"Moubarek, Master. Halim Gonios (""William George"")",male,NaN,1,1,2661,15.2458,NaN,C
439,440,0,2,"Kvillner, Mr. Johan Henrik Johannesson",male,31.0,0,0,C.A. 18723,10.5000,NaN,S
840,841,0,3,"Alhomaki, Mr. Ilmari Rudolf",male,20.0,0,0,SOTON/O2 3101287,7.9250,NaN,S
720,721,1,2,"Harper, Miss. Annie Jessie ""Nina""",female,6.0,0,1,248727,33.0000,NaN,S
39,40,1,3,"Nicola-Yarred, Miss. Jamila",female,14.0,1,0,2651,11.2417,NaN,C


## 10. Dataset Information

### Concept

The `info()` function provides a summary of the DataFrame structure.

It shows the number of entries, column names, non-null values, and data types.

### Example

For the Titanic dataset, `info()` can help identify whether columns such as `Age` or `Cabin` contain missing values.

### Implementation

The `info()` method is used to display a complete structural summary.

### Output

The output shows the number of rows, non-null counts, and data types.

### Interpretation

Differences between the total number of rows and non-null values may indicate missing data.

### Insight

This is one of the fastest ways to understand the overall structure and quality of a dataset.

### AI/ML Relevance

Missing values and incorrect data types must be identified before preprocessing and model training.

In [11]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 118.9 KB


## 11. Statistical Summary using describe()

### Concept

The `describe()` function provides descriptive statistics for numerical columns.

It can show values such as count, mean, standard deviation, minimum, maximum, and quartiles.

### Example

For the `Age` column, descriptive statistics can help us understand the average age and range of Titanic passengers.

### Implementation

The `describe()` function is applied to the dataset.

### Output

The output displays statistical information for numerical columns.

### Interpretation

The statistics provide a quick summary of the distribution and spread of numerical variables.

### Insight

Different numerical columns may have different ranges and scales.

### AI/ML Relevance

Statistical summaries help identify unusual values, possible outliers, and potential requirements for feature scaling.

In [12]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


## 12. Inspecting Data Types using dtypes

### Concept

The `dtypes` attribute shows the specific data type of every column.

### Example

Numerical columns may use integer or floating-point data types, while text-based columns may use object data types.

### Implementation

We use `df.dtypes` to inspect all column data types.

### Output

The output lists each column and its corresponding data type.

### Interpretation

The result helps classify columns into numerical and categorical groups.

### Insight

Data type inspection is necessary before selecting appropriate analysis and preprocessing methods.

### AI/ML Relevance

Numerical and categorical variables require different preprocessing techniques before they can be used in many Machine Learning models.

In [13]:
print(df.dtypes)

PassengerId      int64
Survived         int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Cabin              str
Embarked           str
dtype: object


## 13. Unique Values using nunique()

### Concept

The `nunique()` function counts the number of unique values in each column.

### Example

The `Sex` column may contain a small number of unique categories, while `Name` may contain many unique values.

### Implementation

We use `nunique()` to count unique values in every column.

### Output

The output displays the number of distinct values for each variable.

### Interpretation

Columns with a small number of unique values may be categorical, while columns with very high uniqueness may be identifiers or high-cardinality features.

### Insight

Unique value counts help us understand the complexity of each variable.

### AI/ML Relevance

High-cardinality columns may require special preprocessing and may not always be suitable for direct model input.

In [14]:
df.nunique()

PassengerId    891
Survived         2
Pclass           3
Name           891
Sex              2
Age             88
SibSp            7
Parch            7
Ticket         681
Fare           248
Cabin          147
Embarked         3
dtype: int64

## 14. Frequency Analysis using value_counts()

### Concept

The `value_counts()` function counts how frequently each unique value appears in a column.

### Example

For the `Sex` column, it can show how many passengers were male and female.

### Implementation

We use `value_counts()` on selected categorical columns.

### Output

The output displays the frequency of each category.

### Interpretation

The result helps us understand the distribution and balance of categorical variables.

### Insight

Some categories may occur much more frequently than others.

### AI/ML Relevance

Understanding category frequencies helps identify class imbalance and categorical distribution before Machine Learning.

In [15]:
print("Sex:")
print(df["Sex"].value_counts())

print("\nPclass:")
print(df["Pclass"].value_counts())

print("\nEmbarked:")
print(df["Embarked"].value_counts())

print("\nSurvived:")
print(df["Survived"].value_counts())

Sex:
Sex
male      577
female    314
Name: count, dtype: int64

Pclass:
Pclass
3    491
1    216
2    184
Name: count, dtype: int64

Embarked:
Embarked
S    644
C    168
Q     77
Name: count, dtype: int64

Survived:
Survived
0    549
1    342
Name: count, dtype: int64


## 15. Numerical Columns

### Concept

Numerical columns contain values that can be measured, counted, or used in mathematical calculations.

### Example

In the Titanic dataset, columns such as `Age`, `Fare`, and `SibSp` contain numerical values.

### Implementation

We use `select_dtypes(include="number")` to identify numerical columns.

### Output

The output displays all columns containing numerical data.

### Interpretation

These variables can be used for statistical analysis and numerical visualizations.

### Insight

Numerical features may require analysis for missing values, distributions, and outliers.

### AI/ML Relevance

Numerical features are commonly used as model inputs and may require scaling depending on the algorithm.

In [16]:
numerical_columns = df.select_dtypes(include="number").columns.tolist()

print("Numerical Columns:")
print(numerical_columns)

Numerical Columns:
['PassengerId', 'Survived', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare']


## 16. Categorical Columns

### Concept

Categorical columns contain labels or groups rather than continuous numerical measurements.

### Example

In the Titanic dataset, `Sex` and `Embarked` represent categories.

### Implementation

We use `select_dtypes(include="object")` to identify text-based categorical columns.

### Output

The output displays columns with object data types.

### Interpretation

These columns may need to be encoded before being used in many Machine Learning models.

### Insight

Some categorical columns may contain a small number of categories, while others may have many unique values.

### AI/ML Relevance

Categorical features often require encoding methods such as one-hot encoding or label encoding.

In [18]:
categorical_columns = df.select_dtypes(include=["object", "string"]).columns.tolist()

print("Categorical Columns:")
print(categorical_columns)

Categorical Columns:
['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked']


## 17. Date/Time Columns

### Concept

Date/time columns store information related to dates, times, or timestamps.

### Example

A business dataset might contain columns such as order date, joining date, or transaction time.

### Implementation

We inspect the Titanic dataset to identify columns stored with date/time data types.

### Output

The output displays any date/time columns found in the dataset.

### Interpretation

If the result is empty, the dataset does not currently contain columns stored as date/time variables.

### Insight

The Titanic dataset is focused on passenger information and does not require date/time information for this inspection.

### AI/ML Relevance

Date/time variables can be useful for feature engineering in datasets involving time-based patterns.

In [19]:
datetime_columns = df.select_dtypes(include=["datetime64[ns]", "datetimetz"]).columns.tolist()

print("Date/Time Columns:")
print(datetime_columns)

Date/Time Columns:
[]


## 18. Identifier Columns

### Concept

An identifier column is used to uniquely identify individual records.

### Example

In the Titanic dataset, `PassengerId` represents a unique identifier for each passenger.

### Implementation

We inspect the number of unique values in `PassengerId`.

### Output

The output compares the number of unique passenger IDs with the total number of rows.

### Interpretation

If the number of unique values is equal to the number of rows, the column can be considered an identifier.

### Insight

Identifier columns are useful for record identification but may not contain meaningful predictive information.

### AI/ML Relevance

Identifier columns are generally excluded from Machine Learning features unless they contain useful information.

In [20]:
print("Total Rows:", len(df))
print("Unique PassengerId Values:", df["PassengerId"].nunique())

if df["PassengerId"].nunique() == len(df):
    print("PassengerId is an identifier column.")

Total Rows: 891
Unique PassengerId Values: 891
PassengerId is an identifier column.


## 19. Potential Target Variable

### Concept

A target variable is the output that a Machine Learning model attempts to predict.

### Example

In the Titanic dataset, `Survived` indicates whether a passenger survived or did not survive.

- 0 represents not survived.
- 1 represents survived.

### Implementation

We inspect the unique values and frequency distribution of the `Survived` column.

### Output

The output displays the number of passengers in each survival category.

### Interpretation

Since `Survived` contains two outcome classes, it can be used as the target variable for a binary classification problem.

### Insight

The dataset can potentially be used to study which passenger characteristics are associated with survival.

### AI/ML Relevance

Features such as passenger class, sex, age, fare, and family-related variables can potentially be used to predict the `Survived` target.

In [21]:
print("Potential Target Variable: Survived")

print("\nUnique Values:")
print(df["Survived"].unique())

print("\nValue Counts:")
print(df["Survived"].value_counts())

Potential Target Variable: Survived

Unique Values:
[0 1]

Value Counts:
Survived
0    549
1    342
Name: count, dtype: int64


# Final Data Inspection Summary

The Titanic dataset has been successfully loaded and inspected.

During this notebook, the following aspects were examined:

- Number of rows
- Number of columns
- Dataset dimensions
- Column names
- Dataset index
- Data types
- First records using `head()`
- Last records using `tail()`
- Random records using `sample()`
- Dataset information using `info()`
- Statistical summary using `describe()`
- Unique values using `nunique()`
- Category frequencies using `value_counts()`
- Numerical columns
- Categorical columns
- Date/time columns
- Identifier columns
- Potential target variable

The inspection shows that the Titanic dataset contains a combination of numerical and categorical variables. `PassengerId` acts as an identifier, while `Survived` is a potential target variable for a binary classification problem.

This initial inspection provides a foundation for the next stages of Exploratory Data Analysis, including data quality assessment, missing-value analysis, duplicate analysis, univariate analysis, and further Machine Learning preparation.